DOCSP-43425 Add sort data (#131)
mongodb/docs-rust: source/fundamentals/crud/read-operations/sort.txt

documents in the ``books`` collection:

.. literalinclude:: /includes/fundamentals/code-snippets/crud/sort.rs
   :start-after: start-book-struct
   :end-before: end-book-struct
   :language: rust
   :dedent:

The following code shows how to insert sample data into the ``books``
collection:

.. literalinclude:: /includes/fundamentals/code-snippets/crud/sort.rs
   :start-after: start-sample-data
   :end-before: end-sample-data
   :language: rust
   :dedent:

Methods for Sorting
-------------------

You can sort results retrieved by a query, or you can sort results within an
aggregation pipeline.

To sort your query results, use the ``sort()`` method of the ``FindOptions``
struct. To sort your results within an aggregation pipeline, create a ``$sort``
stage and pass the list of stages to the ``aggregate()`` method.

Use the ``sort()`` method of the ``FindOptions`` struct to sort documents in
ascending order of the values of the ``author`` fields as shown in the following
code:

.. literalinclude:: /includes/fundamentals/code-snippets/crud/sort.rs
   :start-after: start-sort-query
   :end-before: end-sort-query
   :language: rust
   :dedent:

> Create a ``$sort`` stage that sorts documents in ascending order by the values
> of the ``author`` field as shown in the following code:

Reviewer: ```suggestion
Alternatively, you can create a ``$sort`` stage that sorts documents in ascending order by the values
of the ``author`` field as shown in the following code:
```

``Sort()`` tells MongoDB to order returned documents by the values of one or more
fields in a certain direction. To sort returned documents by a field in
ascending (lowest first) order, use a value of ``1``. To sort in descending
(greatest first) order instead, use ``-1``. If you do not specify a sort,
MongoDB does not guarantee the order of query results.

Sample Data for Examples
------------------------

The examples in this guide use the following ``Book`` struct as a model for
documents in the ``books`` collection:

.. literalinclude:: /includes/fundamentals/code-snippets/crud/sort.rs
   :start-after: start-book-struct
   :end-before: end-book-struct
   :language: rust
   :dedent:

The following code shows how to insert sample data into the ``books``
collection:

.. literalinclude:: /includes/fundamentals/code-snippets/crud/sort.rs
   :start-after: start-sample-data
   :end-before: end-sample-data
   :language: rust
   :dedent:

Methods for Sorting
-------------------

You can sort results retrieved by a query, or you can sort results within an
aggregation pipeline.

To sort your query results, use the ``sort()`` method of the ``FindOptions``

> struct. To sort your results within an aggregation pipeline, use the ``$sort``
> stage of the ``aggregate()`` method.

Reviewer: ```suggestion
struct. To sort your results within an aggregation pipeline, create a ``$sort``
stage and pass the list of stages to the ``aggregate()`` method.
```

> You can use the ``sort()`` method of the ``FindOptions`` struct to sort
> documents by ``author`` as follows:

Reviewer: ```suggestion
Use the ``sort()`` method of the ``FindOptions`` struct to sort
documents in ascending order of values of the ``author`` field as shown in the following code:
```

.. literalinclude:: /includes/fundamentals/code-snippets/crud/sort.rs
   :start-after: start-sort-query
   :end-before: end-sort-query
   :language: rust
   :dedent:

> You can use the ``$sort`` stage of the ``aggregate()`` method within an
> aggregation pipeline to sort documents by ``author`` as follows:

Reviewer: ```suggestion
Create a ``$sort`` stage that sorts documents in ascending order by the values of the ``author``
field as shown in the following code:
```

.. literalinclude:: /includes/fundamentals/code-snippets/crud/sort.rs
   :start-after: start-sort-aggregation
   :end-before: end-sort-aggregation
   :language: rust
   :dedent:

Sorting Direction
-----------------

The direction of your sort can either be **ascending** or **descending**. An
ascending sort orders your results from smallest to largest. A descending sort
orders your results from largest to smallest.

> Here are some examples of data sorted in ascending order:

Reviewer: ```suggestion
The following list contains examples of data sorted in ascending order:
```

* Numbers: 1, 2, 3, 43, 43, 55, 120
* Dates: 1990-03-10, 1995-01-01, 2005-10-30, 2005-12-21 
* Words (ASCII): Banana, Dill, carrot, cucumber, hummus

> Here are some examples of data sorted in descending order:

Reviewer: ```suggestion
The following list contains examples of data sorted in descending order:
```

* Numbers: 100, 30, 12, 12, 9, 3, 1
* Dates: 2020-01-01, 1998-12-11, 1998-12-10, 1975-07-22 
* Words (reverse ASCII): pear, grapes, apple, Cheese

The following subsections show how to specify these sort criteria.

> Ascending
> ~~~~~~~~~

Reviewer: ```suggestion
Ascending Sort
~~~~~~~~~~~~~~
```

Author: Hmm I see what you mean, although other driver sort data pages just use `Ascending` so will opt to stick with that for consistency 👍

> An ascending sort orders your results from smallest to largest. To specify this sort,
> pass the field you want to sort by and ``1`` to the ``sort()`` method within the
> ``FindOptions`` struct.

Reviewer: S: remove redundant info

```suggestion
To specify an ascending sort,
pass the field you want to sort by and ``1`` to the ``sort()`` method of the
``FindOptions`` type.
```

Example
^^^^^^^

The following example specifies an ascending sort on the ``name`` field:

.. io-code-block::
   :copyable: true

   .. input:: /includes/fundamentals/code-snippets/crud/sort.rs
      :start-after: start-ascending-sort
      :end-before: end-ascending-sort
      :language: rust
      :dedent:

   .. output::
      :language: console
      :visible: false

      { "_id": 6, "name": "A Dance with Dragons", "author": Martin, "length": 1104 }
      { "_id": 3, "name": "Atlas Shrugged", "author": Rand, "length": 1088 }
      { "_id": 5, "name": "Crypotnomicon", "author": Stephenson, "length": 918 }
      { "_id": 4, "name": "Infinite Jest", "author": Wallace, "length": 1104 }
      { "_id": 2, "name": "Les Miserables", "author": Hugo, "length": 1462 }

>       { "_id": 1, "name": "The Brothers Karamazov", "author": Dostoevsky, "length": 824 }

Reviewer: S: investigate why these aren't being printed as Book types. Maybe you can use this api to print the cursor instead:
```rust

let mut cursor = my_coll
--
.find(doc! { "$and": vec!
[
doc! { "unit_price": doc! { "$lt": 12.00 } },
doc! { "category": doc! { "$ne": "kitchen" } }
] })
.sort(doc! { "unit_price": -1 })
.await?;
 
while let Some(result) = cursor.try_next().await? {
println!("{:?}", result);
}

```

Author: Yes this one worked better! Also cleaner to read I think. Updating!

  
Descending
~~~~~~~~~~

> A descending sort orders your results from largest to smallest. To specify this
> sort, pass the field you want to sort by and ``-1`` to the ``sort()`` method
> within the ``FindOptions`` struct.

Reviewer: ```suggestion
To specify a descending
sort, pass the field you want to sort by and ``-1`` to the ``sort()`` method
of the ``FindOptions`` type.
```

Example
^^^^^^^

The following example specifies a descending sort on the ``name`` field:

.. io-code-block::
   :copyable: true

   .. input:: /includes/fundamentals/code-snippets/crud/sort.rs
      :start-after: start-descending-sort
      :end-before: end-descending-sort
      :language: rust
      :dedent:

   .. output::
      :language: console
      :visible: false

      { "_id": 1, "name": "The Brothers Karamazov", "author": Dostoevsky, "length": 824 }
      { "_id": 2, "name": "Les Miserables", "author": Hugo, "length": 1462 }
      { "_id": 4, "name": "Infinite Jest", "author": Wallace, "length": 1104 }
      { "_id": 5, "name": "Crypotnomicon", "author": Stephenson, "length": 918 }
      { "_id": 3, "name": "Atlas Shrugged", "author": Rand, "length": 1088 }
      { "_id": 6, "name": "A Dance with Dragons", "author": Martin, "length": 1104 }

Additional Information
----------------------

To learn more about the operations metnioned in this guide, see the following:

> - :ref:`rust-query-guide`

Reviewer: S: add links in other guides to this sort guide

Reviewer: I was also thinking that you could add links to this page in the options [here](https://deploy-preview-131--docs-rust.netlify.app/fundamentals/crud/read-operations/retrieve/#:~:text=Default%3A%20None-,sort,-The%20sort%20to) and [here](https://deploy-preview-131--docs-rust.netlify.app/fundamentals/aggregation/#additional-information)
maybe also [here](https://deploy-preview-131--docs-rust.netlify.app/fundamentals/crud/read-operations/text-search/#sort-by-relevance)

You can sort results retrieved by a query, or you can sort results within an
aggregation pipeline.

Chain the ``sort()`` method to the ``find()`` method to sort results retrieved
by the query, as shown in the following example:

.. literalinclude:: /includes/fundamentals/code-snippets/crud/sort.rs
   :start-after: start-sort-query
   :end-before: end-sort-query
   :language: rust
   :dedent:

Multiple Options
~~~~~~~~~~~~~~~~

Alternatively, you can use the ``sort()`` method of the ``FindOptions``

> struct. This is useful for configuring multiple options. 

Reviewer 2: ```suggestion
struct.
```

I suggest omitting the last sentence and rephrasing the title on line 70. We expect that the main use case for the `with_options` method is caching/reusing options across multiple method calls, so that the list of options doesn't need to be re-written. This is more useful when there are more options being configured, but it's not _necessarily_ preferable to construct an options struct when multiple options are set. Sorry, my original comment wasn't very clear on this distinction!

Author: Ah ok that makes sense! Will update. For the ``Ascending`` and ``Descending`` examples, would you suggest updating to use the fluent-style API there as well as opposed to using the ``with_options`` method as they are now?

Reviewer 2: Yep, I think the fluent API is better-suited for these examples

Author: Sounds good. Just updated 👍

Author: @isabelatkinson If the latest changes look good to you, can you ✅ when you have a moment? Ty!
